pkg/serializer/internal/metrics/iterable_series.go

```diff
@@ -263,165 +368,154 @@ func (series *IterableSeries) MarshalSplitCompress(bufferContext *marshaler.Buff
+			return ps.Embedded(serieMetadataOrigin, func(ps *molecule.ProtoStream) error {
+				if serie.NoIndex {
+					err = ps.Int32(serieMetadataOriginMetricType, metryTypeNotIndexed)
 					if err != nil {
 						return err
 					}
-					return ps.Int32(serieMetadataOriginOriginService, metricSourceToOriginService(serie.Source))
-				})
+				}
+				err = ps.Int32(serieMetadataOriginOriginProduct, serieMetadataOriginOriginProductAgentType)
+				if err != nil {
+					return err
+				}
+				err = ps.Int32(serieMetadataOriginOriginCategory, metricSourceToOriginCategory(serie.Source))
+				if err != nil {
+					return err
+				}
+				return ps.Int32(serieMetadataOriginOriginService, metricSourceToOriginService(serie.Source))
 			})
-			if err != nil {
-				return err
-			}
-			return nil
 		})
 		if err != nil {
-			return nil, err
+			return err
 		}
+		return nil
+	})
+	if err != nil {
+		return err
+	}
 
-		if len(serie.Points) > maxPointsPerPayload {
-			// this series is just too big to fit in a payload (even alone)
-			err = stream.ErrItemTooBig
-		} else if pointsThisPayload+len(serie.Points) > maxPointsPerPayload {
-			// this series won't fit in this payload, but will fit in the next
-			err = stream.ErrPayloadFull
-		} else {
-			// Compress the protobuf metadata and the marshaled series
-			err = addToPayload()
-		}
+	if len(serie.Points) > pb.maxPointsPerPayload {
+		// this series is just too big to fit in a payload (even alone)
+		err = stream.ErrItemTooBig
+	} else if pb.pointsThisPayload+len(serie.Points) > pb.maxPointsPerPayload {
+		// this series won't fit in this payload, but will fit in the next
+		err = stream.ErrPayloadFull
+	} else {
+		// Compress the protobuf metadata and the marshaled series
+		err = addToPayload()
+	}
 
-		switch err {
-		case stream.ErrPayloadFull:
-			expvarsPayloadFull.Add(1)
-			tlmPayloadFull.Inc()
+	switch err {
+	case stream.ErrPayloadFull:
+		expvarsPayloadFull.Add(1)
+		tlmPayloadFull.Inc()
 
-			err = finishPayload()
-			if err != nil {
-				return nil, err
-			}
+		err = pb.finishPayload()
+		if err != nil {
+			return err
+		}
 
-			err = startPayload()
-			if err != nil {
-				return nil, err
-			}
+		err = pb.startPayload()
+		if err != nil {
+			return err
+		}
 
-			// Add it to the new compression buffer
-			err = addToPayload()
-			if err == stream.ErrItemTooBig {
-				// Item was too big, drop it
-				expvarsItemTooBig.Add(1)
-				tlmItemTooBig.Inc()
-				continue
-			}
-			if err != nil {
-				// Unexpected error bail out
-				expvarsUnexpectedItemDrops.Add(1)
-				tlmUnexpectedItemDrops.Inc()
-				return nil, err
-			}
-		case stream.ErrItemTooBig:
+		// Add it to the new compression buffer
+		err = addToPayload()
+		if err == stream.ErrItemTooBig {
 			// Item was too big, drop it
 			expvarsItemTooBig.Add(1)
-			tlmItemTooBig.Add(1)
-		case nil:
-			continue
-		default:
+			tlmItemTooBig.Inc()
+			// TODO: was continue
```

Comment:
Can you comment this `return nil`, something along this line:

```
// we tried making it fit in the original payload, but there is no way
// it can goes in here, it will be dropped: instrument the error and report nil
// since there is nothing we can do here.
```

TBH you probably know better than me what's tried here, my suggestion might be wrong 😅 


You also have a `TODO` line 495 that you can either remove or slightly comment (despite the overall design being awful, this second `return nil` is easier to understand).

Reply:
Good catch, replaced the TODOs and cleaned up the comments.